### ISSUE TITLE
v7.0.0 ARM64 binaries don't work on Ubuntu 18.04

### ISSUE THREAD

**Issue by aknuds1** [grafana/grafana#24844]
Grafana v7.0.0 ARM64 binaries don't work on Ubuntu 18.04, due to a dynamic linking failure. They do work on Ubuntu 20.04 however.

Launching grafana-server fails as follows:

```
grafana-server[13734]: /usr/sbin/grafana-server: /lib/aarch64-linux-gnu/libc.so.6: version `GLIBC_2.28’ not found (required by /usr/sbin/grafana-server)
```

Output of ldd --version:

```
ldd (Ubuntu GLIBC 2.27-3ubuntu1) 2.27
```

I suggest we try building on Ubuntu 18.04, to see if that produces a binary that also works on 20.04. We should also try Debian oldstable (suggested by @RichiH).

I've made [a PR](https://github.com/grafana/grafana/pull/24967) to switch our Docker build image to Debian Stretch - it can also easily be switched to use Ubuntu 18.04 instead, depending on what we want to use.

**Comment by aknuds1** [comment]
I can confirm it doesn't work on Ubuntu 18.04, but Ubuntu 20.04 is fine. Must be a binary compatibility issue.

**Comment by damm** [comment]
I don't run Grafana in Docker I rather wish I did as I am resorting to rebuilding Grafana from source to get it running because of this issue.  Previous builds were excellent, and I don't have a 20.04 upgrade path yet for my ARM devices.

**Comment by aknuds1** [comment]
Yeah sorry :/ We didn't realize this issue would arise, we're looking into ways to solve it!

**Comment by damm** [comment]
Yeah let me know when that happens and I'll upgrade to 7; my build from source didn't work.

**Comment by RobertMochnacky** [comment]
Hi there,
I have the same problem after the upgrade from Grafana 6.7 to 7.0. I'm running it on Raspbian OS (Debian strech 9 distro).
Can you please advise how to fix that?
Thank you.

**Comment by aknuds1** [comment]
@RobertMochnacky I'm sorry, but I'm not sure how to resolve this binary compatibility issue on Ubuntu 18.04. The only solution that I can see myself is to compile Grafana from source.

We are trying to find a solution to this binary portability problem though, but I can't promise any time frame.

**Comment by aMU5Ed** [comment]
> Hi there,
> I have the same problem after the upgrade from Grafana 6.7 to 7.0. I'm running it on "Raspbian GNU/Linux 9 (stretch) on Raspberry Pi 3.
> Can you please advise how to fix that?
> Thank you.
> 
> grafana-server[4475]: /usr/sbin/grafana-server: /lib/arm-linux-gnueabihf/libc.so.6: version `GLIBC_2.28' not found (required by /usr/sbin/grafana-server)

Hi, what's the output of `ldd --version` ?

**Comment by torkelo** [comment]
What Ubuntu version do we use when we build the arm binaries?

**Comment by aknuds1** [comment]
@torkelo It's Ubuntu 20.04. I've made an [issue](https://github.com/grafana/grafana/issues/24844) to try using Ubuntu 18.04 instead, or potentially Debian stable.

**Comment by RobertMochnacky** [comment]
Hello,

the output of ldd --version is as follows:

ldd (Debian GLIBC 2.24-11+deb9u4) 2.24
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.

**Comment by torkelo** [comment]
@aknuds1 do we build binaries on ubuntu 20? that seems wrong, we usually need to build binaries on very old linux versions to have good binary compatibility, one of the reasons why https://github.com/grafana/grafana-build-container/blob/master/Dockerfile  is based on ubuntu 14 is to have good GLIBC Compatability

**Comment by aknuds1** [comment]
@torkelo Afraid so, [this](https://github.com/grafana/grafana/blob/master/scripts/build/ci-build/Dockerfile) is the Dockerfile for the build image.

I don't think we can use a _too_ old version of Ubuntu either though, since they go out of support. My current thinking is to test Ubuntu 18.04 and Debian stable.

**Comment by massimilianocasini** [comment]
Hi there, same problem of @aMU5Ed on Raspbian (GNU/Linux 9 (stretch) , after upgrade to 7.0

ldd (Debian GLIBC 2.24-11+deb9u4) 2.24

grafana-server: /lib/arm-linux-gnueabihf/libc.so.6: version `GLIBC_2.28' not found (required by grafana-server)

**Comment by damm** [comment]
Targetting Raspbian would work better than Ubuntu 18.04.  Not everyone can upgrade their distro from 18.04 to 18.10 at the snap of their fingers.

https://twitter.com/armbian/status/1259435342412865537 sounds nice, but not everyone runs Armbian.

**Comment by aknuds1** [comment]
> Targetting Raspbian would work better than Ubuntu 18.04. Not everyone can upgrade their distro from 18.04 to 18.10 at the snap of their fingers.

@damm What are you suggesting exactly? That we build on another distro than Ubuntu 18.04?

**Comment by aknuds1** [comment]
As an update to this issue, I've tested a build image based on Debian Stretch, it produces binaries compatible with Ubuntu 18.04 (Debian Buster does not).

**Comment by damm** [comment]
@aknuds1 Indeed I was suggesting to not use Ubuntu.  Debian Stretch seems like a solid choice.

as I am aware of several organizations that still run 16.04 in production which might seem crazy however Ubuntu is still shipping backports until April of 2021.

**Comment by aknuds1** [comment]
@damm Gotcha. Can you outline why you think Debian is a better choice than Ubuntu in this regard? I'm trying to collect pros/cons.

**Comment by hrw** [comment]
Distro | libc6 version
-|-
Ubuntu xenial | 2.23
Debian stretch| 2.24
Ubuntu bionic|2.27
Debian buster|2.28
Ubuntu focal | 2.30

Xenial is still supported but most people moved to Bionic. Stretch is no longer supported by Debian.

Ubuntu bionic (18.04) looks best still.

**Comment by damm** [comment]
@aknuds1 well based on the table that @hrw listed it sounds like 16.04 would be a better bet.  I thought that Debian would be more conservative with its Glibc version than Ubuntu would be.

**Comment by abrownsword** [comment]
I am also having the original post's problem.  I am currently running Ubuntu 18.04, so a binary built on that or 16.04 should be fine (or Debian stretch).  I tried to build Grafana 7 from source, but ran into issues doing so as well and don't really want to unravel the complexities of the Go environment on ARM64 in order to move to Grafana 7.

**Comment by RichiH** [comment]
Oldstable and oldoldstable still get security updates, and "support" does
not matter much on a build image.

Sent by mobile; please excuse my brevity.

**Comment by aknuds1** [comment]
I've switched our build image to be based on Debian Stretch - this should fix the binary incompatibilities with older Linuxes. The fix should be included with the next official Grafana release.

**Comment by drpixel972** [comment]
Hello,
The same issue is present for CentOS 7 aarch64. 
Best Regards,

**Comment by aknuds1** [comment]
@drpixel972 This issue has been solved with v7.0.3. Have you tested that version?

**Comment by drpixel972** [comment]
My bad, I didn't notice that 7.03 was out ... 
It's working indeed :-)
Thank you
